### DISCUSSION TITLE
Amount of "Sends"

### DISCUSSION THREAD

**Discussion by markusschloesser** [markusschloesser/MackieC4_P3#39] [Ideas]
I didn't know this, so I tried it. The maximum amount of "Sends" that Live allows is 12.
That's why Leigh probably implemented exactly that amount.
As I have never used more than 5, I'm gonna reduce those from 12 to 8, in order to make room for other functionality (unfold group track). 
It might be a bad idea though, because then a banking feature for Sends would need to be implemented.
Will try

**Comment by BilldarBagdar** [comment]
There's no harm in simply declaring we're only going to support the first 8 sends on any track.  How many people use more than 8 sends in Live AND have a C4?  LOL, they aren't our "target demographic"... :)

Although a thought I've been kicking around might apply. there are 4 theoretical "modes" currently, but there are also 4 modifier keys that are not used yet, which means the true theoretical limit should be 16 modes.  We could have, for example, Channel Strip, Shift + Channel Strip, and Alt + Channel Strip modes, where most of the controls are mapped the same in all 3 modes. Like Pan, Volume, and the device row.  But then, for example, one of those modes has the 12 send controls, and another has 0, 2, 4, or whatever send controls and a bunch of other stuff mapped to those 12, 10, 8 other formerly-send encoders.

**Comment by BilldarBagdar** [comment]
```
Python: INFO:MarkusC4test.MackieC4:726 - (MackieC4) Param name <A-Reverb>
Python: INFO:MarkusC4test.MackieC4:726 - (MackieC4) Param name <B-Delay>
```
Above is what lands in the logs from this code below when a "real send" exists  
```
        if vpot_index < len(self.song().view.selected_track.mixer_device.sends):
            p = self.song().view.selected_track.mixer_device.sends[vpot_index]
            self.main_script().log_message("Param name <{0}>".format(p.name))
            return (p, p.name)
        else:
            # The Song doesn't have this many sends
            return None, '      '  # remove this text after you see it in the LCD, just use blanks
```
 So this is the None you are seeing for all the unmapped sends.  Why are you seeing it?  Because instead of being 'p' a send parameter it is None a None object.  The code must not be checking for None before simply printing what is expected to be 'p' but is actually None to the LCD.

I pushed an update to the implementLiveObjectValid branch.  You should no longer see None in the LCDs for Sends and Device parameters that don't exist.  You will see something else that should be easier to find...  :)

**Reply by markusschloesser** [reply]
since I only had 1 Espresso until now, I don't get it how the OOOMMMM (you should have reversed this ;-) )helps.
but fyi: I updated the "implementLiveObjectValid" branch to unlight the vpots when disconnecting, this also now helps when initializing because there are no (still) enlightened vpots, meaning the state of the vpots reflects the state of Live.
BTW do you have an idea where the "there are always 12 sends" is defined? I was looking for that yesterday but then got distracted by the not working destroy function (and half fixed it).

**Comment by BilldarBagdar** [comment]
:)  oooo mmmm can now easily be removed, just do "Find" on "oooo".  I could have left it as blanks myself, but then it would have been harder for you to find and review what was done to fix the issue.

The code quoted above puts None in place for every "send encoder" index where there is not actually a send on the selected Track.  The "send encoders" are "defined" [here](https://github.com/markusschloesser/MackieC4_P3/blob/implementLiveObjectValid/wip/MackieC4/EncoderController.py#L869) in `__reassign_encoder_parameters()`  (same place to fix OOOMMM)

`elif s_index in row_01_encoders:` means the encoders 8 - 16, so the next condition `elif s_index < encoder_29_index:` will never see an index less than the encoder 16 index (which is 15), and so the `<` condition means the send encoders are encoders 17 - 28.

So, if one wanted to "free-up" the last 4 send encoders for other functions in Channel Strip mode, you could do this:
```
if s_index in row_00_encoders:
    # no changes
elif s_index in row_01_encoders:
    # no changes
elif s_index in row_02_encoders: 
    # move the send encoder stuff here
elif s_index < encoder_29_index:
    # these encoders (25, 26, 27, 28) are now free for other use
else:
    # no changes
```

Of course, this would be just the first step, right?  Once you make this change, you would need to go through the code and update all the places where encoders (25, 26, 27, 28) are referenced for behavior/functionality, like what happens when the encoder button is pushed for example.

**Reply by markusschloesser** [reply]
"As I have never used more than 5, I'm gonna reduce those from 12 to 8, in order to make room for other functionality (unfold group track)." 
I decided to leave the sends alone for now quantity-wise, as they are in a too messed up state right now, instead I am trying to fix sends and more importantly: will try to implement group track fold/unfold on encoder 4 in chan strip mode (btw I think we should switch those modes rather sooner than later, because they are wrongly done/reversed).

**Reply by BilldarBagdar** [reply]
> I think we should switch those modes rather sooner than later

Do you mean the "Assignment box" (C4 graphics) buttons and how they are mapped to "script behavior" modes?

Unfortunately, Marker, Track, Function, and "Chan Strip" do not intuitively map well for our needs.  Also, "Chan Strip" and "Track" kind of mean the same thing to me in Live terms.  (some DAWs "always" have EQ controls, for example, in a "channel strip" view, and their "Track view" wouldn't show any EQ unless you specifically loaded an EQ plugin; In Live, there isn't really a concept of "channel strip EQ", everything like that is a "plugin" (a device))

Currently, the button-to-mode mapping is:

- Marker =  User mode
- Track = device mode
- Function = song mode
- Chan Strip = track mode

Correct?  What mapping would you propose? 

Maybe we should think about putting together a "design document" on the wiki where we can talk about, plan, and document decisions related to stuff like this.  Right now, with only 2 or 3 functional modes, it's pretty easy to keep everything straight, so now is a great time to start such documentation.  Imagine what such a document could grow into though if we implement 4 modes and 4 "modified" versions of each mode, and how easy it should be to figure out, for example, "What does the "Spot Erase" or "Split" button do in all 20 modes?" (4 modes + (4 * 4) modified modes)

**Reply by markusschloesser** [reply]
> Do you mean the "Assignment box" (C4 graphics) buttons and how they are mapped to "script behavior" modes?

Yes 

> Unfortunately, Marker, Track, Function, and "Chan Strip" do not intuitively map well for our needs. Also, "Chan Strip" and "Track" kind of mean the same thing to me in Live terms. (some DAWs "always" have EQ controls, for example, in a "channel strip" view, and their "Track view" wouldn't show any EQ unless you specifically loaded an EQ plugin; In Live, there isn't really a concept of "channel strip EQ", everything like that is a "plugin" (a device))

For me "track" is Volume, Pan, Mute, Solo (and now fold/unfold ;-) ), and "chan strip" is what's ON the track. SO I suggest to just switch those two around.


> 
> Currently, the button-to-mode mapping is:
> 
> * Marker =  User mode
> * Track = device mode
> * Function = song mode
> * Chan Strip = track mode
> 
> Correct? What mapping would you propose?
> 
> Maybe we should think about putting together a "design document" on the wiki where we can talk about, plan, and document decisions related to stuff like this. Right now, with only 2 or 3 functional modes, it's pretty easy to keep everything straight, so now is a great time to start such documentation. Imagine what such a document could grow into though if we implement 4 modes and 4 "modified" versions of each mode, and how easy it should be to figure out, for example, "What does the "Spot Erase" or "Split" button do in all 20 modes?" (4 modes + (4 * 4) modified modes)

While I like the idea of a design document, that's what I've been using the Issues=enhancement for. Don't you think we can use that?

**Reply by BilldarBagdar** [reply]
Yes and no.  I mean "issues" are like "problems with the implementation of the design" and sometimes you just need to know, "what did we want this chingadero to do?  It's doing one thing that is a problem, but what was it supposed to do?" A "problem" would be an "issue" proper... But the "design document" would have the answers to the "what was it supposed to do?" kinds of questions. 

A design document becomes like the "blueprints" and "renderings" associated with an architect's "project design" for a building.  As coders on this project we are both Architect and General Contractor, so we don't need to hand-off our design to the builders (with enough details that we can blame them for "not following the plans" if things go south) and we can avoid a lot of the "design specification" because we are the builders and the designers. (and because we don't have a team of corporate forensic analysts on the hunt for the root cause of every "production outage" related to our application ..... [double rabbit hole: I am pretty brutal with "Microsoft presets" I am allowed to remove from my "Windows experience".  If I can, if it lets me, I remove it, first thing after Windows starts for the first time.  and I only put back (let back) stuff I find I need and want on that particular machine. {triple rabbit hole: I've got 4 "music studio machines" that will never be storing "3-D objects", (DRM) "Music" or "Saved Games", but I can't get rid of those from the Explorer. It really irritates me that MS puts that crap front and center in Explorer with no way to make it disappear if you aren't going to use the "feature", or simply want to see something else front and center.} That is to say, Windows key + '.' just types ..... for me, no emojis, on this box.]) But even so, it wouldn't be a bad idea to lay out and document the plan, such as it may be.  Even most of the DIY house-builders out there at least have blueprints to build from.

Something standard, but any standard would need to develop on its own, like for now, for example, going forward we could document that we've decided on implementing these 4 buttons mapped this way:

- Marker = User mode
- Track = track mode
- Function = song mode
- Chan Strip = device mode

and then for example, possibly declare that the buttons across the bottom in the "Function", Assignment", and "Modifiers" "graphic boxes" have "global functionality" across all modes; whereas the 4 buttons around the "black ellipse" and the 4 inside the "Parameter box" have distinct functionality within each selected mode.  Then maybe we would, for example, state what each button is supposed to do in each mode, or across all modes.  Then, when we are down, elbows deep in the code and have a question like, for example, "What behavior is mapped to the "Lock" button click in "Option + Function" mode?"; or, "What behavior is mapped to "encoder 32 button click + hold" in "Alt + User" mode?" you have a place you can go look up the answer, or add your answer in the case when you "create" functionality because "it wasn't mapped already" or whatever.

and of course, this is 2021 so nothing is chiseled in stone.  If we later decide that the "function box" buttons will have special "function mode" behavior only, or behavior mapped that is unique in every mode (as opposed to global behavior).  We just make the design change decision, and then update the document.  

I can personally attest to the time saving and head-scratch saving and disagreement settling qualities a "just good enough" design document can provide.  I can also attest to the overblown bureaucratic size and relative uselessness of "poor" (IMHO) design documentation.

It would also seem fair to me to categorize such a "design documentation" effort as double dipping into the "user help/how to documentation" bucket.  If we've been keeping track of that stuff ourselves anyway, then copy/paste into the "getting started guide" and "advanced modes" documents becomes a breeze.
